<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DREAMS is an analysis pipeline that applies a neural network to low-frequency variant calling and circulating tumor DNA detection from next-generation DNA sequencing data.">
<title>Deep Read-level Error Model for Sequencing data • dreams</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Deep Read-level Error Model for Sequencing data">
<meta property="og:description" content="DREAMS is an analysis pipeline that applies a neural network to low-frequency variant calling and circulating tumor DNA detection from next-generation DNA sequencing data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">dreams</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="index.html">
    <span class="fa fa-solid fa-cloud-moon"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JakobSkouPedersenLab/dreams/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="dreams">DREAMS<a class="anchor" aria-label="anchor" href="#dreams"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>DREAMS is an analysis pipeline that applies a neural network to low-frequency variant calling and circulating tumor DNA detection from next-generation DNA sequencing data.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of dreams from <a href="https://github.com/JakobSkouPedersenLab/dreams" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"JakobSkouPedersenLab/dreams"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="additional-setup-if-needed">Additional Setup (If needed)<a class="anchor" aria-label="anchor" href="#additional-setup-if-needed"></a>
</h3>
<p>If you encounter any issues related to TensorFlow integrations within R, install Keras within the correct python environment to ensure a proper setup:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">keras</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/keras/man/install_keras.html" class="external-link">install_keras</a></span><span class="op">(</span>envname <span class="op">=</span> <span class="st">"&lt;ENVIRONMENT_NAME&gt;"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>After installation, set the environment at the start of each R session:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_condaenv</a></span><span class="op">(</span><span class="st">"&lt;ENVIRONMENT_NAME&gt;"</span>, required <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="basic-functions">Basic Functions<a class="anchor" aria-label="anchor" href="#basic-functions"></a>
</h2>
<p>This section provides an overview of the basic functions available in the dreams library.</p>
<div class="section level3">
<h3 id="data-preparation-and-model-training">Data Preparation and Model Training<a class="anchor" aria-label="anchor" href="#data-preparation-and-model-training"></a>
</h3>
<p>The first steps in using DREAMS involve preparing your data and setting up the model for training, crucial for effective variant calling and cancer detection.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/JakobSkouPedersenLab/dreams" class="external-link">dreams</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For training, DREAMS requires one or more BAM files and a reference genome.</span></span>
<span><span class="va">training_data</span> <span class="op">=</span> <span class="fu"><a href="reference/get_training_data.html">get_training_data</a></span><span class="op">(</span></span>
<span>  bam_paths <span class="op">=</span> <span class="st">"/path/bam_file"</span>,</span>
<span>  reference_path <span class="op">=</span> <span class="st">"/path/hg38.fa"</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Training the DREAMS Model using a Neural Network</span></span>
<span><span class="co"># Basic settings for Keras are required.</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">=</span> <span class="fu"><a href="reference/train_dreams_model.html">train_dreams_model</a></span><span class="op">(</span></span>
<span>  <span class="va">training_data</span>,</span>
<span>  layers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">128</span>, <span class="fl">64</span>, <span class="fl">32</span><span class="op">)</span>,</span>
<span>  model_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"read_index"</span>, <span class="st">"strand"</span>, <span class="st">"trinucleotide_ctx"</span>, <span class="st">"first_in_pair"</span>, </span>
<span>                     <span class="st">"umi_count"</span>, <span class="st">"seq_length"</span>, <span class="st">"fragment_size"</span>, <span class="st">"n_other_errors"</span>, </span>
<span>                     <span class="st">"local_GC"</span><span class="op">)</span>,</span>
<span>  lr <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">32000</span>,</span>
<span>  epochs <span class="op">=</span> <span class="fl">750</span>,</span>
<span>  model_file_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="feature-selection">Feature Selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h4>
<p>The DREAMS model supports a variety of features categorized into numeric, categorical, and embedded types:</p>
<div class="section level5">
<h5 id="numeric-features">Numeric Features<a class="anchor" aria-label="anchor" href="#numeric-features"></a>
</h5>
<ul>
<li>
<code>read_index</code>, <code>fragment_size</code>, <code>local_GC</code>, <code>umi_count</code>, <code>umi_errors</code>, <code>local_complexity_1</code>, <code>local_complexity_2</code>, <code>n_other_errors</code>, <code>prior_error</code>, <code>seq_length</code>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="categorical-features">Categorical Features<a class="anchor" aria-label="anchor" href="#categorical-features"></a>
</h5>
<ul>
<li>
<code>ref</code>, <code>strand</code>, <code>first_in_pair</code>, <code>ctx_minus1</code>, <code>ctx_plus1</code>, <code>chr</code>, <code>genomic_pos</code>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="embedded-feature">Embedded Feature<a class="anchor" aria-label="anchor" href="#embedded-feature"></a>
</h5>
<ul>
<li><code>trinucleotide_ctx</code></li>
</ul>
<p>Ensure that the dataset used aligns with the selected features and adjust the parameters such as <code>layers</code>, <code>lr</code>, <code>batch_size</code>, and <code>epochs</code> as needed.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="variant-calling-and-cancer-detection-with-dreams">Variant Calling and Cancer Detection with DREAMS<a class="anchor" aria-label="anchor" href="#variant-calling-and-cancer-detection-with-dreams"></a>
</h3>
<p>The statistical methods <code>dreams_vc</code> and <code>dreams_cc</code> can be used for variant calling and cancer detection, respectively.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Call variants using DREAMS-vc</span></span>
<span></span>
<span><span class="va">variant_calls</span> <span class="op">=</span> <span class="fu"><a href="reference/dreams_vc.html">dreams_vc</a></span><span class="op">(</span></span>
<span>  mutations_df <span class="op">=</span> <span class="va">mutations_df</span>,</span>
<span>  bam_file_path <span class="op">=</span> <span class="st">"/path/test_bam_file"</span>,</span>
<span>  reference_path <span class="op">=</span> <span class="st">"/path/hg38.fa"</span>,</span>
<span>  model <span class="op">=</span> <span class="va">model</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Call cancer using DREAMS-cc</span></span>
<span></span>
<span><span class="va">cancer_calls</span> <span class="op">=</span> <span class="fu"><a href="reference/dreams_cc.html">dreams_cc</a></span><span class="op">(</span></span>
<span>  mutations_df <span class="op">=</span> <span class="va">mutations_df</span>,</span>
<span>  bam_file_path <span class="op">=</span> <span class="st">"/path/test_bam_file"</span>,</span>
<span>  reference_path <span class="op">=</span> <span class="st">"/path/hg38.fa"</span>,</span>
<span>  model <span class="op">=</span> <span class="va">model</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parallel-variant-calling-with-dreams_vc_parallel">Parallel Variant Calling with dreams_vc_parallel<a class="anchor" aria-label="anchor" href="#parallel-variant-calling-with-dreams_vc_parallel"></a>
</h3>
<p>In addition to <code>dreams_vc</code>, the DREAMS package offers <code>dreams_vc_parallel</code> for parallel variant calling. This function is particularly useful when working with large datasets and is designed to leverage computational resources such as multi-core CPUs for parallel processing.</p>
<div class="section level4">
<h4 id="when-to-use-dreams_vc_parallel">When to use dreams_vc_parallel<a class="anchor" aria-label="anchor" href="#when-to-use-dreams_vc_parallel"></a>
</h4>
<ul>
<li>
<strong>Large Datasets</strong>: Efficiently handles larger datasets by distributing the workload across multiple cores.</li>
<li>
<strong>Parallel Processing Capability</strong>: Ideal when the computational environment supports parallel processing (e.g., multi-core systems).</li>
</ul>
</div>
<div class="section level4">
<h4 id="example-usage">Example Usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Parallel variant calling</span></span>
<span><span class="va">parallel_variant_calls</span> <span class="op">=</span> <span class="fu"><a href="reference/dreams_vc_parallel.html">dreams_vc_parallel</a></span><span class="op">(</span></span>
<span>  mutations_df <span class="op">=</span> <span class="va">mutations_df</span>,</span>
<span>  bam_file_path <span class="op">=</span> <span class="st">"/path/test_bam_file"</span>,</span>
<span>  reference_path <span class="op">=</span> <span class="st">"/path/hg38.fa"</span>,</span>
<span>  model <span class="op">=</span> <span class="va">model</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="saving-and-loading-models">Saving and Loading Models<a class="anchor" aria-label="anchor" href="#saving-and-loading-models"></a>
</h3>
<p>You can save your trained models for later use and load them as needed. To save a trained model, use the <code>save_model_hdf5</code> function from the <code>keras</code> package. Specify the file path where you want to save the model. Similarly, when training a model using the <code>train_dreams_model</code> function, you can directly specify a file path where to save the model using the <code>model_file_path</code> argument. This allows for automatic saving of the model upon training completion.</p>
<p>As default <code>model_file_path = NULL</code>, and the model won’t be saved automatically. You can then manually save the model using <code>save_model_hdf5</code>.</p>
<p>To load a previously saved model, use the <code>load_model_hdf5</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tensorflow.rstudio.com/" class="external-link">keras</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save the model</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/keras/man/save_model_hdf5.html" class="external-link">save_model_hdf5</a></span><span class="op">(</span><span class="va">model</span>, filepath <span class="op">=</span> <span class="st">"path/to/your_model.h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the model</span></span>
<span><span class="va">loaded_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/save_model_hdf5.html" class="external-link">load_model_hdf5</a></span><span class="op">(</span>filepath <span class="op">=</span> <span class="st">"path/to/your_model.h5"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="about-dreams">About DREAMS<a class="anchor" aria-label="anchor" href="#about-dreams"></a>
</h2>
<p>For technical details describing how DREAMS works please see our <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02920-1" class="external-link">article</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/JakobSkouPedersenLab/dreams/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/JakobSkouPedersenLab/dreams/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing dreams</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Mikkel Hovden Christensen <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2588-3036" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Simon Opstrup Drue <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1401-6527" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/JakobSkouPedersenLab/dreams/actions" class="external-link"><img src="https://github.com/JakobSkouPedersenLab/dreams/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://zenodo.org/badge/latestdoi/455089263" class="external-link"><img src="https://zenodo.org/badge/455089263.svg" alt="DOI"></a></li>
<li><a href="https://app.codecov.io/gh/JakobSkouPedersenLab/dreams?branch=main" class="external-link"><img src="https://codecov.io/gh/JakobSkouPedersenLab/dreams/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://www.gnu.org/licenses/gpl-3.0" class="external-link"><img src="https://img.shields.io/badge/License-GPLv3-blue.svg" alt="License: GPL v3"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Mikkel Hovden Christensen, Simon Opstrup Drue.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
